### DISCUSSION TITLE
RDF*/SPARQL* support?

### DISCUSSION THREAD

**Discussion by sa-bpelakh** [RDFLib/rdflib#1554] [Ideas]
There is currently a W3C working group considering the adoption of Olaf Hartig's work on RDF*/SPARQL* as a W3C standard (http://olafhartig.de/files/Hartig_AMW2017_RDFStar.pdf). Several DB vendors already support this syntax (Anzo, Stardog, Blazegraph), and others are working on adding it as a feature. It provides a great bridge between RDF and Property graphs.

Could this be possibly added to `rdflib`?

**Comment by gromgull** [comment]
Since RDFLib has essentially 0 active developers and we don't even manage
to make a release - I don't see that as very likely :D

Unless someone contributes a very good PR !

- Gunnar

On Thu, 6 Feb 2020 at 20:02, Boris Pelakh <notifications@github.com> wrote:

> There is currently a W3C working group considering the adoption of Olaf
> Hartig's work on RDF*/SPARQL* as a W3C standard (
> http://olafhartig.de/files/Hartig_AMW2017_RDFStar.pdf). Several DB
> vendors already support this syntax (Anzo, Stardog, Blazegraph), and others
> are working on adding it as a feature. It provides a great bridge between
> RDF and Property graphs.
>
> Could this be possibly added to rdflib?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RDFLib/rdflib/issues/955?email_source=notifications&email_token=AABLO3F3OTAZ3534CIEUQFDRBRNEVA5CNFSM4KRCIRJKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4ILTLQTQ>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABLO3CSUSLR4UK3V5Y5CMTRBRNEVANCNFSM4KRCIRJA>
> .
>


-- 
http://gromgull.net

**Comment by JervenBolleman** [comment]
@gromgull I am working on it! Hope to publish a branch next week.

**Comment by sa-bpelakh** [comment]
@JervenBolleman that is fantastic,  please let me know if you need a hand with testing.

**Comment by JervenBolleman** [comment]
The approach I am taking is by going the RDF* PG mode where <ex:p ex:y ex:z> asserts the triple and itself is a special BNode (basically a bnode with a derivable _:id) Which maps to reified statements. Has it's downsides but I think making RDF* a syntactic sugar is the way to go. At least with my limited skills.

**Comment by JervenBolleman** [comment]
Hi, just wanted to note that my slow progress is [available here](https://github.com/JervenBolleman/rdflib/tree/rdfstar1). I am however, having difficulty with the parser code. So I was very over optimistic on getting this work faster.

**Comment by JervenBolleman** [comment]
I am now into adding the new EmpTP nodes into the algebra. Recursive `<<?s ?o <<?s2 ?o2 ?p2>>>>` does not work, as I don't know how to define that in the parser logic.

**Comment by JervenBolleman** [comment]
Well "it works" now. PG mode, SPARQL only by translating the sparql* into a reification quad.
Needs serious work to polish it all up and make it more Pythonic.

**Comment by nicholascar** [comment]
Great work here @JervenBolleman! We (my company) will likely be looking at RDF* soon for W3C graph work (GQL).

**Comment by JervenBolleman** [comment]
@nicholascar thank you. Still a lot of work to be done. 

- [x] Move EmbTP to reification quad translation into evaluation from algebra (Via new triple object)
- [ ] Implement SPARQL* update
- [ ] Make SPARQL* recursive in the parser
- [ ] Add many more test cases!
- [ ] BIND and VALUES need to work as well.
- [x] Figure out why URIRef inside the `<<>>` are not transformed into URIRef objects

**Comment by JervenBolleman** [comment]
Just want to share my progress. The parsing is done better now, and instead of hacking in support to the algebra it is now pushed down into the evaluation. However, I have not yet managed to get the evaluate to work nicely. It is natural to put it in the evalBGP part, but resolving an "EmbeddedTriple" to a series of Identifiers is quite painful there. I need to think a bit more.

**Comment by pragya16067** [comment]
Hi @JervenBolleman, I along with a group of students from IIIT Delhi ardently wish to contribute to this issue. We would like to get in contact with you soon if possible and would be very obliged if you could walk us through the work done so far. 
Looking forward to your response!

**Comment by JervenBolleman** [comment]
@pragya16067 find my e-mail and contact me.

**Comment by pragya16067** [comment]
Hey @JervenBolleman, mailed!

**Comment by nicholascar** [comment]
Position paper on RDF*/SPARQL* at https://blog.liu.se/olafhartig/2019/01/10/position-statement-rdf-star-and-sparql-star/

**Comment by JervenBolleman** [comment]
[This preprint has a lot of details required for implementing](https://arxiv.org/abs/1406.3399). Includes the required changes to the grammar.

**Comment by pragya16067** [comment]
Thanks @JervenBolleman and @nicholascar. Will take a look!

**Comment by nicholascar** [comment]
Consider the alternate Turtle syntax proposal by Stardog (they seem to have implemented it already): https://www.stardog.com/blog/property-graphs-meet-stardog/

**Comment by hartig** [comment]
Just for the record: another triple store that supports RDF* and SPARQL* now is Ontotext GraphDB (as of version 9.2); see http://graphdb.ontotext.com/documentation/9.2/free/devhub/rdf-sparql-star.html

The main part of their implementation is within their GraphDB code base but some of the base stuff is in the underlying [RDF4J library](https://rdf4j.org/) they are using internally.

Another Java library that you may look at to get inspiration for adding RDF*/SPARQL* support in RDFLib is [Apache Jena](https://jena.apache.org/) and the [RDFTools library](https://github.com/RDFstar/RDFstarTools) we have built on top if it. At the core, the object model of Jena has been extended with a class called [Node_Triple](http://jena.apache.org/documentation/javadoc/jena/index.html?org/apache/jena/graph/Node_Triple.html) that can be used in the [Triple class](http://jena.apache.org/documentation/javadoc/jena/index.html?org/apache/jena/graph/Triple.html). Also, there are grammars for Turtle* and for the SPARQL* syntax.

**Comment by aarushiag** [comment]
Hi @JervenBolleman and @nicholascar , we have tried to make some changes for making the SPARQL* recursive in nature. Kindly review it  :) (PR Link - https://github.com/JervenBolleman/rdflib/pull/1 )

**Comment by pragya16067** [comment]
Hi @JervenBolleman and @nicholascar, I have added support for parsing and querying on RDF* graphs via some additional wrapper functions in the RDF (Turtle) parser. 
Kindly review it :) (PR Link - https://github.com/JervenBolleman/rdflib/pull/4 )

**Comment by pragya16067** [comment]
> Great work here @JervenBolleman! We (my company) will likely be looking at RDF* soon for W3C graph work (GQL).

Hi @nicholascar. I have added a PR (https://github.com/RDFLib/rdflib/pull/1111) with my work on adding support for parsing RDF* graphs. Kindly review it :) Hope to hear from you soon.

**Comment by alex-randles** [comment]
Hi Guys,

I am curious to know is there any update on this?

Thanks!

**Comment by nicholascar** [comment]
Hi @alex-randles JervenBolleman reviewed the PR associated with this issue, #1111, so that's the latest direct work.

The other maintainers and I are still performing a bunch of clean-up updates to the code base before an RDFlib 6.0.0 release - mainly reducing the number of small PRs and tidying up old functions - before we tackle this major enhancement. So we will linkey only get to directly focussing on this work in a couple of months however it would be great if you or others could review and comment on PR #1111. I think it will need updating and a bit of work to pass tests still, if the approach is deemed appropriate.

**Comment by JervenBolleman** [comment]
Hi All, part of the work is also waiting for the SPARQL* specifications to
be more solid.
Especially in regards to the BIND operator, which currently needs a lot of
work.
This is because BIND for triple patterns is quite different from the
current BIND implementation, and this is where I have not continued my work
at all.
i.e. BIND is a subselect/filter more than a function call/rename.

I think that it would also be very good to modularize the Query parsing so
that it is explicit if a Query is SPARQLstar or SPARQL11.



On Wed, Sep 23, 2020 at 12:25 AM Nicholas Car <notifications@github.com>
wrote:

> Hi @alex-randles <https://github.com/alex-randles> JervenBolleman
> reviewed the PR associated with this issue, #1111
> <https://github.com/RDFLib/rdflib/pull/1111>, so that's the latest direct
> work.
>
> The other maintainers and I are still performing a bunch of clean-up
> updates to the code base before an RDFlib 6.0.0 release - mainly reducing
> the number of small PRs and tidying up old functions - before we tackle
> this major enhancement. So we will linkey only get to directly focussing on
> this work in a couple of months however it would be great if you or others
> could review and comment on PR #1111
> <https://github.com/RDFLib/rdflib/pull/1111>. I think it will need
> updating and a bit of work to pass tests still, if the approach is deemed
> appropriate.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RDFLib/rdflib/issues/955#issuecomment-697014376>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHQYFJR3AOEBFGDTVWXUP3SHEP33ANCNFSM4KRCIRJA>
> .
>


-- 
Jerven Bolleman
me@jerven.eu

**Comment by westurner** [comment]
Are there rdf-star issues and/or mailing list posts for the considered blocking SPARQL* spec issue(s)?

## RDF*
RDF*: RDF*, SPARQL*, N-Triples*, Turtle*
- Web: https://w3c.github.io/rdf-star/
- Src: https://github.com/w3c/rdf-star
- Issues: https://github.com/w3c/rdf-star/issues
- MailingList: https://lists.w3.org/Archives/Public/public-rdf-star/
- UseCases: https://w3c.github.io/rdf-star/UCR/rdf-star-ucr.html
- Spec: RDF*: https://w3c.github.io/rdf-star/rdf-star-cg-spec.html
- Spec: SPARQL*: https://w3c.github.io/rdf-star/tests/sparql/syntax/manifest.html
- Spec: RDF* Semantics: https://w3c.github.io/rdf-star/tests/semantics/manifest.html
- Spec: Turtle*: https://w3c.github.io/rdf-star/tests/turtle/syntax/manifest.html
- Spec: N-Triples*: https://w3c.github.io/rdf-star/tests/nt/syntax/manifest.html


```markdown
## RDF*
RDF*: RDF*, SPARQL*, N-Triples*, Turtle*
- Web: https://w3c.github.io/rdf-star/
- Src: https://github.com/w3c/rdf-star
- Issues: https://github.com/w3c/rdf-star/issues
- MailingList: https://lists.w3.org/Archives/Public/public-rdf-star/
- UseCases: https://w3c.github.io/rdf-star/UCR/rdf-star-ucr.html
- Spec: RDF*: https://w3c.github.io/rdf-star/rdf-star-cg-spec.html
- Spec: SPARQL*: https://w3c.github.io/rdf-star/tests/sparql/syntax/manifest.html
- Spec: RDF* Semantics: https://w3c.github.io/rdf-star/tests/semantics/manifest.html
- Spec: Turtle*: https://w3c.github.io/rdf-star/tests/turtle/syntax/manifest.html
- Spec: N-Triples*: https://w3c.github.io/rdf-star/tests/nt/syntax/manifest.html
```

**Comment by Crazytieguy** [comment]
Hi! Is there any progress on this?

**Comment by hartig** [comment]
The work on the RDF-star & SPARQL-star spec has been completed:
https://www.w3.org/2021/12/rdf-star.html

**Comment by nicholascar** [comment]
Great news @hartig!

I’ve just advertised within my University for students to work on RDFLib’s RDF-Star implementation. Hopefully we get a student starting on this in February. 

The student will pick up on whatever’s implemented, even up to benchmarking a complete implementation against other systems, if we see a full implementation soon.

So if anyone reading this wants to contribute to this implementation, please go ahead and don’t want for students - we may not get any after all. No efforts will be wasted.

**Comment by niklasl** [comment]
This is good. Just remember that any such feature shouldn't be merged into RDFLib main until this becomes a W3C Rec (and the WG hasn't been formed yet). Things can still change from a CG _report_ to a WG _recommentation_ (just look at JSON-LD 1.1) and we don't want deployments depending on a pre-WG-implementation. (Of course, a Rec would need independent implementations, to ratify it.)

**Comment by westurner** [comment]
So, an actual implementation of the actual spec (as-is, with e.g. a SemVer
alpha version) should be an rdflib plugin until W3C Rec status?

Rdflib plugins are discovered by their setup tools entry_points metadata:
https://rdflib.readthedocs.io/en/stable/plugins.html
https://github.com/RDFLib/rdflib-jsonld/blob/master/setup.py#L104-L111

What needs to be done to get RDF*/SPARQL* support for rdflib in an rdflib/
org repo?

- [ ] parser
- [ ] serializer

What else?

- [ ] modify/fork existing non RDF*/SPARQL* store plugins?
- [ ] modify/fork existing non RDF*/SPARQL* query result plugins?

**Reply by mwatts15** [reply]
Regarding a new SPARQL* parser plugin, [I suggested before](https://github.com/RDFLib/rdflib/issues/1406#issuecomment-927190512) copying the SPARQL parser to create a such a plugin rather than modifying the existing SPARQL parser. I haven't created a SPARQL* parser this way because I'm not actually in favor of RDF\*/SPARQL\*, but I *did* come up with decent way of copying the parser while still having access to all of the names in the module where the parser is defined.

I defined an object with references to sub-parsers here: https://github.com/RDFLib/rdflib/compare/master...mwatts15:copyable-sparql-parser?expand=1
You can copy `PARSER_REFS` defined in there for your customized parser.

**Comment by sdasda7777** [comment]
Hi, it appears the forks mentioned are inactive and no longer developed. Are there any recent efforts for RDF*/SPARQL* implementation?

**Comment by niklasl** [comment]
Hi @sdasda7777! RDF-star (as it is now called, to aid in searching etc.), is currently undergoing standardization in the [W3C RDF-star Working Group](https://www.w3.org/groups/wg/rdf-star/). of which I'm a member.

While outside of the RDFLib codebase, I have two Python projects, [TRLD](https://github.com/niklasl/trld) and [RQLD](https://github.com/niklasl/rqld), of which the former has support for parsing and serializing the RDF-star annotation syntax (to the still very much moving target of JSON-LD-star; users be warned). The latter, parsing SPARQL, will have support for SPARQL-star once more details have been settled (my efforts in this area currently goes into what's being specified, so implementation is mostly on hold). In one way or another I aim to bring what's in that to RDFLib too, time permitting (at worst through a thin adapter).

There's a lot going on in the details (regarding exactly what is supposed to be added to RDF 1.2), but I have hopes that we're converging on something good now (from having been worried about complexity, risk of tripping up users, etc.).
